### PR TITLE
Logging out of Gitlab and clearing session storage

### PR DESCRIPTION
**Pull request by atomicgamedeveloper**
Continuing off of commit ddab8e97030bc1a2b2ea06310d252df3a78c2ead of pull request https://github.com/INTO-CPS-Association/DTaaS/pull/681, this PR aims to fix issue https://github.com/INTO-CPS-Association/DTaaS/issues/654 and https://github.com/INTO-CPS-Association/DTaaS/issues/479. This commit fixes an invalid hook call in the `MenuToolbar` component by passing its `AuthContextProps` object to the `signOut` function. This keeps the hook call top-level as per the rules of hooks.

### CODE REVIEW

File: client/test/unitTests/Util/Auth/Authentication.test.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { signOut } from 'util/auth/Authentication';
+import { useAuth } from 'react-oidc-context';
+
+jest.mock('react-oidc-context');
```
Comment (prasadtalasila) [inline client/test/unitTests/Util/Auth/Authentication.test.ts:4]: This https://github.com/INTO-CPS-Association/DTaaS/blob/feature/distributed-demo/client/test/integration/authRedux.test.tsx#L10 test might be useful

**Comment by prasadtalasila** [comment]
@atomicgamedeveloper thanks for the update.

I see an error with the following command.

```bash
$yarn clean install
$yarn build
yarn run v1.22.22
warning ../../../../../package.json: No license field
$ npx react-scripts build
/home/dtaas/.npm/_npx/cd53aabf71f3f793/node_modules/resolve/lib/sync.js:113
    throw err;
    ^

Error: Cannot find module 'typescript' from '/home/dtaas/Documents/git/into-cps/DTaaS/client/node_modules'
    at Function.resolveSync [as sync] (/home/dtaas/.npm/_npx/cd53aabf71f3f793/node_modules/resolve/lib/sync.js:111:15)
    at getModules (/home/dtaas/.npm/_npx/cd53aabf71f3f793/node_modules/react-scripts/config/modules.js:119:32)
    at Object.<anonymous> (/home/dtaas/.npm/_npx/cd53aabf71f3f793/node_modules/react-scripts/config/modules.js:142:18)
    at Module._compile (node:internal/modules/cjs/loader:1369:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1427:10)
    at Module.load (node:internal/modules/cjs/loader:1206:32)
    at Module._load (node:internal/modules/cjs/loader:1022:12)
    at Module.require (node:internal/modules/cjs/loader:1231:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (/home/dtaas/.npm/_npx/cd53aabf71f3f793/node_modules/react-scripts/config/webpack.config.js:28:17) {
  code: 'MODULE_NOT_FOUND'
}

Node.js v20.12.2
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
This is an error due to outdated dependencies. We need to resolve this error before the code can be build. Do you know a fix? Thanks.

**Comment by prasadtalasila** [comment]
My nodejs stack is

```bash
$yarn -v
1.22.22
$node -v
v20.12.2
$npm -v
10.5.0
```

**Comment by atomicgamedeveloper** [comment]
Mine is:
yarn -v
1.22.22
node -v
v20.14.0
npm -v
10.7.0

I can try downgrading node and npm and remaking it from there.

**Comment by prasadtalasila** [comment]
@atomicgamedeveloper I am not able to replicate the logout functionality.

What I've tried:

yarn install
yarn build
yarn config:test
yarn start

Accessed the app at localhost:4000 and logged in using gitlab.com credentials. I've also used self-hosted gitlab solution. I am not able to logout in both cases.

**Comment by prasadtalasila** [comment]
@atomicgamedeveloper please also rebase your branch with the latest commit on `feature/distributed-demo` branch. Thanks.

**Comment by prasadtalasila** [comment]
I've also checked with docker compose combination in `compose.dev.yml` and the logout does not happen there either. Please let me know if I'm doing something wrong. Thanks.

**Comment by atomicgamedeveloper** [comment]
@prasadtalasila The PR is ready for review, thank you.

I have also looked into how long fetch could potentially run for in the current implementation: By default, [Fetch API has no timeout set](https://betterstack.com/community/guides/scaling-nodejs/nodejs-timeouts/#timing-out-a-fetch-api-request) so the request may hang forever, but the [AbortSignal.timeout()](https://nodejs.org/api/globals.html#static-method-abortsignaltimeoutdelay) API provides an easy way to implement it.

**Comment by prasadtalasila** [comment]
@atomicgamedeveloper please use AbortSignal with a timeout of 30 seconds.

**Comment by prasadtalasila** [comment]
I see two pending issues.

### Unpredictable URL

```js
await fetch(`${window.env.REACT_APP_URL}_oauth/logout`, {
      signal: AbortSignal.timeout(30000),
    });
```

Here `${window.env.REACT_APP_URL}_oauth` might resolve either to `${window.env.REACT_APP_URL}_oauth` or `${window.env.REACT_APP_URL}/_oauth` depending on how the `REACT_APP_URL` is specified in `env.js`. It is better to use `useAppURL()` from `util/envUtils.ts` to always get a URL without `/` at the end. Then you can always use `${window.env.REACT_APP_URL}/_oauth`. Sorry that I didn't see this earlier.

### Unseen Call

When I run the logout in client website, I don't see `window.env.REACT_APP_URL}_oauth/logout` call. The call is made to `window.env.REACT_APP_URL}/logout`. Please see the attached screen recording. 

https://github.com/INTO-CPS-Association/DTaaS/assets/9206466/3db07d1c-6202-4413-84e9-3ea404d395c5

Do you know why?

## Indentation

Please set your VS Code indentation to two spaces. Or run `yarn format`. This command will correct the indentation problem. Please remember to run at least

```sh
yarn install
yarn syntax
yarn format
yarn build
yarn config:<select>
yarn start
yarn clean
```

**Comment by atomicgamedeveloper** [comment]
> ### Unseen Call
> 
> When I run the logout in client website, I don't see `window.env.REACT_APP_URL}_oauth/logout` call. The call is made to `window.env.REACT_APP_URL}/logout`. Please see the attached screen recording.
> logout_record.mp4
> 
> Do you know why?

I am having trouble recreating this issue. Could you please tell me if I am missing something?

1. I run these commands in the docker directory:
```
docker compose -f compose.dev.yml down;  
docker compose -f compose.dev.yml build;  
docker compose -f compose.dev.yml up -d;
```
2. I goto `localhost` in a fresh private window.
3. I sign into Gitlab and authorize the backend.
4. I click the **Sign In with Gitlab** button and authorize the frontend.

I assume this is where screen recording starts.

5. I run the logout.

From the reload I get redirected to Gitlab to authorize the backend, which I do. It doesn't seem like you need to do this, maybe due to residual cookies?

7. I open a new tab and logout of Gitlab.
8. I stop `docker-traefik-forward-auth-1` with this command:
```
docker stop docker-traefik-forward-auth-1
```
9. I click the **Sign In with Gitlab** button
10. I start `docker-traefik-forward-auth-1` with this command:
```
docker start docker-traefik-forward-auth-1
```
11. I log into Gitlab and authorize the frontend again.
12. I stop `docker-traefik-forward-auth-1` again with this command:
```
docker stop docker-traefik-forward-auth-1
```
13. I run logout.

Here I am redirected to a `404 page not found` page (unlike the blank page in the video) and the following 7 requests. Please also note the revoke requests that are also missing from the video's final network activities.

image

I have added a 3 second timeout before refreshing as I ran into a similar issue after adding it in. Kindly let me know if it changes anything.

**Comment by prasadtalasila** [comment]
The tests are failing,

```sh
$yarn test:unit
yarn run v1.22.22
warning ../../../../../package.json: No license field
$ jest -c ./jest.config.json ../test/unitTests --coverage
 FAIL  test/unitTests/Util/Auth/Authentication.test.ts (12.603 s)
  signOut
    ✕ expires _xsrf cookie (27 ms)
    ✓ does not signoutRedirect if auth.user is null (3 ms)
    ✕ signsOutRedirect, clearStaleState, removeTokens and removeUer if user is authorized (1 ms)
    ✕ fetches the URI from window.env (1 ms)
    ✕ throws an error if fetch rejects (28 ms)
    ✕ throws an error if signoutRedirect rejects (9 ms)
    ✕ reloads the page (1 ms)
    ✕ clears sessionStorage

  ● signOut › expires _xsrf cookie

    TypeError: (0 , envUtil_1.useAppURL) is not a function

      28 |   }
      29 |   const LOGOUT_URL = getLogoutRedirectURI() ?? '';
    > 30 |   const APP_URL = cleanURL(useAppURL());
         |                                     ^
      31 |   const idToken = auth.user.id_token;
      32 |   try {
      33 |     await auth.revokeTokens();

      at src/util/auth/Authentication.ts:30:37
      at src/util/auth/Authentication.ts:1547:39
      at Object.<anonymous>.__awaiter (src/util/auth/Authentication.ts:1496:10)
      at signOut (src/util/auth/Authentication.ts:1634:10)
      at test/unitTests/Util/Auth/Authentication.test.ts:82:18
      at test/unitTests/Util/Auth/Authentication.test.ts:8:71
      at Object.<anonymous>.__awaiter (test/unitTests/Util/Auth/Authentication.test.ts:4:12)
      at Object.<anonymous> (test/unitTests/Util/Auth/Authentication.test.ts:80:41)

  ● signOut › signsOutRedirect, clearStaleState, removeTokens and removeUer if user is authorized

    TypeError: (0 , envUtil_1.useAppURL) is not a function

      28 |   }
      29 |   const LOGOUT_URL = getLogoutRedirectURI() ?? '';
    > 30 |   const APP_URL = cleanURL(useAppURL());
         |                                     ^
      31 |   const idToken = auth.user.id_token;
      32 |   try {
      33 |     await auth.revokeTokens();

      at src/util/auth/Authentication.ts:30:37
      at src/util/auth/Authentication.ts:1547:39
      at Object.<anonymous>.__awaiter (src/util/auth/Authentication.ts:1496:10)
      at signOut (src/util/auth/Authentication.ts:1634:10)
      at test/unitTests/Util/Auth/Authentication.test.ts:101:18
      at test/unitTests/Util/Auth/Authentication.test.ts:8:71
      at Object.<anonymous>.__awaiter (test/unitTests/Util/Auth/Authentication.test.ts:4:12)
      at Object.<anonymous> (test/unitTests/Util/Auth/Authentication.test.ts:99:104)

  ● signOut › fetches the URI from window.env

    TypeError: (0 , envUtil_1.useAppURL) is not a function

      28 |   }
      29 |   const LOGOUT_URL = getLogoutRedirectURI() ?? '';
    > 30 |   const APP_URL = cleanURL(useAppURL());
         |                                     ^
      31 |   const idToken = auth.user.id_token;
      32 |   try {
      33 |     await auth.revokeTokens();

      at src/util/auth/Authentication.ts:30:37
      at src/util/auth/Authentication.ts:1547:39
      at Object.<anonymous>.__awaiter (src/util/auth/Authentication.ts:1496:10)
      at signOut (src/util/auth/Authentication.ts:1634:10)
      at test/unitTests/Util/Auth/Authentication.test.ts:111:18
      at test/unitTests/Util/Auth/Authentication.test.ts:8:71
      at Object.<anonymous>.__awaiter (test/unitTests/Util/Auth/Authentication.test.ts:4:12)
      at Object.<anonymous> (test/unitTests/Util/Auth/Authentication.test.ts:108:52)

  ● signOut › throws an error if fetch rejects

    expect(received).rejects.toThrow(expected)

    Expected message: "Error occurred during logout: foo"
    Received message: "(0 , envUtil_1.useAppURL) is not a function"

          28 |   }
          29 |   const LOGOUT_URL = getLogoutRedirectURI() ?? '';
        > 30 |   const APP_URL = cleanURL(useAppURL());
             |                                     ^
          31 |   const idToken = auth.user.id_token;
          32 |   try {
          33 |     await auth.revokeTokens();

          at src/util/auth/Authentication.ts:30:37
          at src/util/auth/Authentication.ts:1547:39
          at Object.<anonymous>.__awaiter (src/util/auth/Authentication.ts:1496:10)
          at signOut (src/util/auth/Authentication.ts:1634:10)
          at test/unitTests/Util/Auth/Authentication.test.ts:121:25
          at test/unitTests/Util/Auth/Authentication.test.ts:8:71
          at Object.<anonymous>.__awaiter (test/unitTests/Util/Auth/Authentication.test.ts:4:12)
          at Object.<anonymous> (test/unitTests/Util/Auth/Authentication.test.ts:118:53)

      119 |     const auth = useAuth();
      120 |     global.fetch = jest.fn().mockRejectedValueOnce('foo');
    > 121 |     await expect(signOut(auth)).rejects.toThrow(
          |                                         ^
      122 |       Error('Error occurred during logout: foo'),
      123 |     );
      124 |   });

      at Object.toThrow (node_modules/@jest/expect/node_modules/expect/build/index.js:218:22)
      at test/unitTests/Util/Auth/Authentication.test.ts:121:41
      at test/unitTests/Util/Auth/Authentication.test.ts:8:71
      at Object.<anonymous>.__awaiter (test/unitTests/Util/Auth/Authentication.test.ts:4:12)
      at Object.<anonymous> (test/unitTests/Util/Auth/Authentication.test.ts:118:53)

  ● signOut › throws an error if signoutRedirect rejects

    expect(received).rejects.toThrow(expected)

    Expected message: "Error occurred during logout: Error: signoutRedirect rejected"
    Received message: "(0 , envUtil_1.useAppURL) is not a function"

          28 |   }
          29 |   const LOGOUT_URL = getLogoutRedirectURI() ?? '';
        > 30 |   const APP_URL = cleanURL(useAppURL());
             |                                     ^
          31 |   const idToken = auth.user.id_token;
          32 |   try {
          33 |     await auth.revokeTokens();

          at src/util/auth/Authentication.ts:30:37
          at src/util/auth/Authentication.ts:1547:39
          at Object.<anonymous>.__awaiter (src/util/auth/Authentication.ts:1496:10)
          at signOut (src/util/auth/Authentication.ts:1634:10)
          at test/unitTests/Util/Auth/Authentication.test.ts:131:25
          at test/unitTests/Util/Auth/Authentication.test.ts:8:71
          at Object.<anonymous>.__awaiter (test/unitTests/Util/Auth/Authentication.test.ts:4:12)
          at Object.<anonymous> (test/unitTests/Util/Auth/Authentication.test.ts:126:63)

      129 |       .fn()
      130 |       .mockRejectedValueOnce(new Error('signoutRedirect rejected'));
    > 131 |     await expect(signOut(auth)).rejects.toThrow(
          |                                         ^
      132 |       Error('Error occurred during logout: Error: signoutRedirect rejected'),
      133 |     );
      134 |   });

      at Object.toThrow (node_modules/@jest/expect/node_modules/expect/build/index.js:218:22)
      at test/unitTests/Util/Auth/Authentication.test.ts:131:41
      at test/unitTests/Util/Auth/Authentication.test.ts:8:71
      at Object.<anonymous>.__awaiter (test/unitTests/Util/Auth/Authentication.test.ts:4:12)
      at Object.<anonymous> (test/unitTests/Util/Auth/Authentication.test.ts:126:63)

  ● signOut › reloads the page

    TypeError: (0 , envUtil_1.useAppURL) is not a function

      28 |   }
      29 |   const LOGOUT_URL = getLogoutRedirectURI() ?? '';
    > 30 |   const APP_URL = cleanURL(useAppURL());
         |                                     ^
      31 |   const idToken = auth.user.id_token;
      32 |   try {
      33 |     await auth.revokeTokens();

      at src/util/auth/Authentication.ts:30:37
      at src/util/auth/Authentication.ts:1547:39
      at Object.<anonymous>.__awaiter (src/util/auth/Authentication.ts:1496:10)
      at signOut (src/util/auth/Authentication.ts:1634:10)
      at test/unitTests/Util/Auth/Authentication.test.ts:138:18
      at test/unitTests/Util/Auth/Authentication.test.ts:8:71
      at Object.<anonymous>.__awaiter (test/unitTests/Util/Auth/Authentication.test.ts:4:12)
      at Object.<anonymous> (test/unitTests/Util/Auth/Authentication.test.ts:136:37)

  ● signOut › clears sessionStorage

    TypeError: (0 , envUtil_1.useAppURL) is not a function

      28 |   }
      29 |   const LOGOUT_URL = getLogoutRedirectURI() ?? '';
    > 30 |   const APP_URL = cleanURL(useAppURL());
         |                                     ^
      31 |   const idToken = auth.user.id_token;
      32 |   try {
      33 |     await auth.revokeTokens();

      at src/util/auth/Authentication.ts:30:37
      at src/util/auth/Authentication.ts:1547:39
      at Object.<anonymous>.__awaiter (src/util/auth/Authentication.ts:1496:10)
      at signOut (src/util/auth/Authentication.ts:1634:10)
      at test/unitTests/Util/Auth/Authentication.test.ts:145:18
      at test/unitTests/Util/Auth/Authentication.test.ts:8:71
      at Object.<anonymous>.__awaiter (test/unitTests/Util/Auth/Authentication.test.ts:4:12)
      at Object.<anonymous> (test/unitTests/Util/Auth/Authentication.test.ts:143:42)
```

Please check the tests both locally and on github actions of your fork. Thanks.

**Comment by prasadtalasila** [comment]
@atomicgamedeveloper pæease rebase to the latest commit on `feature/distributed-demo` branch. Thanks.

**Pull request merged**